src/machines/java/com/enderio/machines/client/gui/widget/EnergyTextboxWidget.java

```diff
@@ -0,0 +1,104 @@
+package com.enderio.machines.client.gui.widget;
+
+import net.minecraft.client.gui.Font;
+import net.minecraft.client.gui.components.EditBox;
+import net.minecraft.network.chat.Component;
+
+import java.text.DecimalFormat;
+import java.text.NumberFormat;
+import java.util.Locale;
+import java.util.function.Supplier;
+
+public class EnergyTextboxWidget extends EditBox {
+    private final Supplier<Integer> maxEnergy;
+    
+    public EnergyTextboxWidget(Supplier<Integer> maxEnergy, Font font, int x, int y, int width, int height, Component message) {
+        super(font, x, y, width, height, message);
+        this.maxEnergy = maxEnergy;
+        setCanLoseFocus(true);
+    }
+	
+	@Override
+    public void setCanLoseFocus(boolean canLoseFocus) {
+        super.setCanLoseFocus(canLoseFocus);
+    }
+
+    //Input only accepts numerals
+    @Override
+    public boolean charTyped(char codePoint, int modifiers) {
+
+        if (codePoint >= 48 && codePoint <= 57) {
+            int cursorPos = getCursorPosition() + 1;
+            super.charTyped(codePoint, modifiers);
+            setValue(formatEnergy(getValue()));
+            moveCursorTo(cursorPos);
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
+        super.keyPressed(keyCode, scanCode, modifiers);
+
+        if (keyCode == 259) {
+            int cursorPos = getCursorPosition();
+            setValue(formatEnergy(getValue()));
+
+            if (getValue().equals("0")) {
+                moveCursorTo(1);
+            } else {
+                moveCursorTo(cursorPos);
+            }
+        }
+
+        return false;
```

Comment:
I don't think you are supposed to return `false` unless you don't consume the input